### PR TITLE
Command recognition patch

### PR DESCRIPTION
**Pull request by yunlong-wang-cn**
Try to resolve:
1. #225 
2. #220 
3. #209 
4. #223
5. #167

### CODE REVIEW

**Comment by yunlong-wang-cn** [comment]
status update:
1. #225 is done
2. #220 is done

File: marimbabot_speech/src/speech_extraction.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+# since I haven't specified the version of TensorFlow, it seems to install the newest version,
+# so according to official instruction, the environment needs to be configured.
+from precise_runner import PreciseEngine, PreciseRunner, ReadWriteStream
+from audio_common_msgs.msg import AudioDataStamped, AudioData
+from std_msgs.msg import Float32
+import rospy
+import webrtcvad
+from playsound import playsound
+
+VAD_ON = False
```
Comment (Flova) [inline marimbabot_speech/src/speech_extraction.py:13]: Why are these globals and not class fields?

File: marimbabot_speech/src/speech2text.py [hunk 2]
```diff
@@ -1,84 +1,109 @@
...
 
-	def run(self):
-		rospy.spin()
-
 if __name__ == '__main__':
-	rospy.init_node('speech_stt_node', log_level=rospy.DEBUG)
-	speech_recognition = STT()
-	speech_recognition.run()
-
+	# TODO: Finish the command extraction
```
Comment (Flova) [inline marimbabot_speech/src/speech2text.py:104]: Are these TODOs still relevant?

File: marimbabot_speech/src/speech2text.py [hunk 3]
```diff
@@ -1,84 +1,109 @@
...
+	# TODO: consider the voice from other group members, i.e. change the dataset and retrain the model
+	rospy.init_node('speech_recognition_node', log_level=rospy.DEBUG)
```
Comment (Flova) [inline marimbabot_speech/src/speech2text.py:106]: Hardcoded loglevel

File: marimbabot_speech/src/speech2text.py [hunk 4]
```diff
@@ -1,84 +1,109 @@
...
+		self.speech_pub.publish(msg)
 
-	def recognize(self, file_path:str):
-		time_0 = time.time()
-		# load audio and pad/trim it to fit 30 seconds
-		audio = whisper.load_audio(file_path)
-		audio = whisper.pad_or_trim(audio)
+	# To address the way how to unpack the data for whisper
+	def unpack_stream(self, data):
+		return np.array(struct.unpack(f"{int(len(data) / 2)}h", bytes(data)), dtype=float) / 526
```
Comment (Flova) [inline marimbabot_speech/src/speech2text.py:59]: This can be done more efficiently by constructing a numpy dtype that matches the struct and then just call np.frombuffer irc.

Comment (Flova) [reply marimbabot_speech/src/speech2text.py:59]: image

See the speedup and it is also simpler.

Comment (Flova) [reply marimbabot_speech/src/speech2text.py:59]: You probably also don't need the `bytes(data)` cast

File: marimbabot_speech/src/speech2text.py [hunk 5]
```diff
@@ -1,84 +1,109 @@
...
 
-	def generate_prompt(self):
-		commands = get_commands()
-		base_prompt = '''Marimbabot is a marimba playing robot arm. The possible commands include:		
-		'''
-		for command in commands:
-			base_prompt += ''.join(f'{command.strip()}, ')
-		rospy.logdebug(f"Generate prompt as: {base_prompt}.")
-		return base_prompt
+	def warm_up(self):
```
Comment (Flova) [inline marimbabot_speech/src/speech2text.py:29]: Why is this needed? You pad to 30 sec ether way right?

File: marimbabot_speech/package.xml [hunk 6]
```diff
@@ -3,6 +3,7 @@
   <name>marimbabot_speech</name>
   <version>0.0.0</version>
   <description>
+<!--      TODO: modifiy this after update-->
```
Comment (Flova) [inline marimbabot_speech/package.xml:6]: TODO

File: marimbabot_speech/config/kws/precise-engine [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#!/home/wang/anaconda3/envs/py36/bin/python
```
Comment (Flova) [inline marimbabot_speech/config/kws/precise-engine:1]: The shebang seems to be specific to your machine.

**Review by Flova — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 8]
```diff
@@ -20,4 +20,6 @@ transformers[sentencepiece]==4.26.0
 catkin_pkg==0.5.2
 openai-whisper==20230314
 webrtcvad==2.0.10
-sounddevice==0.4.6
\ No newline at end of file
+sounddevice==0.4.6
+precise-runner==0.3.1
```
[the remark this answers is not in the document]
Comment (yunlong-wang-cn) [reply requirements.txt:24]: I haven't test other version, at least this version works.

File: marimbabot_speech/src/speech2text.py (hunk 5 above)
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/speech2text.py:24]: input 30sec blank data to warm up the whisper, since the whisper takes the fixed length of audio data as input.  Or we also can input random length of data, then we must pad or trim it to 30 sec.

File: marimbabot_speech/src/speech2text.py (hunk 4 above)
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/speech2text.py:55]: As I remember, I have tried your solution, and the buffer has not successfully been interpreted. But not pretty sure of it, I will try again and then tell you the result.

File: marimbabot_speech/src/speech2text.py (hunk 3 above)
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/speech2text.py:100]: my bad. done

File: marimbabot_speech/src/text2command.py [hunk 9]
```diff
@@ -1,48 +1,183 @@
...
+		# "preview in 80 bpm",
+		# "preview in forte",
+	]
+	command_extractor = CommandExtraction()
+	command_list = simple_command_list + complicated_command_list
+	for command_text in command_list:
+		command_dict = command_extractor.extract(command_text)
+		print(f"{command_text}:{command_dict}")
+
+def dummy_clinet_command_subscriber():
```
[the remark this answers is not in the document]
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/text2command.py:46]: just for debug, it's deleted.

File: marimbabot_speech/src/text2command.py [hunk 10]
```diff
@@ -1,48 +1,183 @@
...
-		for command in self.command_list:
+
+	def speech_callback(self, req):
+		text = req.text
+		if req.no_speech_prob > 0.5:
+			rospy.logwarn("no_speech_prob > 0.5, but is passed the wad, something weird happened, ignore this speech")
+			return
+		command = self.extract(text)
+		rospy.logdebug(f"command extracted: {command}")
+		command_str = json.dumps(command)
```
[the remark this answers is not in the document]
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/text2command.py:93]: I think the JSON is much more flexible than the message, it's much easier to change when new requirements come in.

[the remark this answers is not in the document]
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/text2command.py:93]: And for interpreting the JSON, just one single line is enough.  `json.dumps(text)` and also for reuse of current msg.

And then, @scaredycode if you are okay with this usage, I will keep it, otherwise we need to talk about a new message format.

File: marimbabot_speech/src/text2command.py [hunk 11]
```diff
@@ -1,48 +1,183 @@
...
+		self.command_pub.publish(msg)
+
+
+	def extract_bpm(self,input_text):
+		pattern = re.compile(r" ([0-9]+) bpm", re.IGNORECASE)
+		result = pattern.findall(input_text)
+		if len(result) > 0:
+			return result[0]
+		else:
+			return ""
```
[the remark this answers is not in the document]
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/text2command.py:106]: `""` doesn't mean an error, just means the parameter is empty, and then the behavior will know, it needs to increase the speed with a certain number of bpm which is hardcoded in the behavior node.

File: README.md [hunk 12]
```diff
@@ -58,6 +58,18 @@ rosdep install serial
 # Install python dependencies
 pip3 install wheel  # firstly install the wheel for further package building if you start from a empty virtual environment.
 pip3 install -r src/marimbabot/requirements.txt
+
+# installation of precise-runner for keyword spooting
+# first, make sure all the necessary software are installed.
+sudo apt install python3-pip \
```
Comment (Juphex) [inline README.md:64]: Please remove this because it is installed via the package.xml / rosdep right now.

**Review by Juphex — COMMENTED** [review]
What is this precise-engine folder about?

File: README.md (hunk 12 above)
Comment (yunlong-wang-cn) [reply README.md:64]: sure

**Comment by yunlong-wang-cn** [comment]
> What is this precise-engine folder about?
@Juphex  contains all the lib for keyword spotting engine, e.i. precise-engine

File: marimbabot_speech/src/speech_extraction.py (hunk 1 above)
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/speech_extraction.py:8]: TODO: put the global variable into constructor.

File: marimbabot_speech/src/speech2text.py (hunk 4 above)
Comment (yunlong-wang-cn) [reply marimbabot_speech/src/speech2text.py]: done

**Comment by yunlong-wang-cn** [comment]
All the command and json pattern refer to there #209  #242

**Comment by yunlong-wang-cn** [comment]
TODO: to solve the precise-engine deployment problem.

**Comment by Juphex** [comment]
@yunlong-wang-cn Is it necessary to have these files inside our code base?

**Comment by yunlong-wang-cn** [comment]
> @yunlong-wang-cn Is it necessary to have these precise engine files inside our code base?

@Juphex I still working on it. It seems works not so well, so I prepared to use abother solution instead use a bunch of the libs

File: marimbabot_speech/src/speech2text.py [hunk 13]
```diff
@@ -1,14 +1,9 @@
-import os
-from nvidia import cudnn
-os.environ['LD_LIBRARY_PATH'] = f'{os.path.dirname(cudnn.__file__)}/lib:$LD_LIBRARY_PATH'
-# since I haven't specified the version of TensorFlow, it seems to install the newest version,
-# so according to official instruction, the environment needs to be configured.
-from marimbabot_msgs.msg import Speech as SpeechMsg
-from audio_common_msgs.msg import AudioData
+import struct
```
Comment (Flova) [inline marimbabot_speech/src/speech2text.py:1]: I am not sure if this is needed anymore

File: marimbabot_speech/src/speech2text.py [hunk 14]
```diff
@@ -56,7 +52,9 @@ def handle_audio_buffer(self, req):
 
 	# To address the way how to unpack the data for whisper
 	def unpack_stream(self, data):
-		return np.array(struct.unpack(f"{int(len(data) / 2)}h", bytes(data)), dtype=float) / 526
+		# return np.array(struct.unpack(f"{int(len(data) / 2)}h", bytes(data)), dtype=float) / 526
+		return np.frombuffer(data, dtype=np.float64)/526
```
Comment (Flova) [inline marimbabot_speech/src/speech2text.py:56]: I think this doesn't work. I casted the bytes to ints and then converted the ints to floats. This changed the binary representation. Loading them directly as floats will result in wrong values.

**Review by Flova — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml [hunk 15]
```diff
@@ -18,6 +18,10 @@ jobs:
     runs-on: ubuntu-20.04
     steps:
     - uses: actions/checkout@v3
+    - name: Install portaudio system dependency
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y portaudio19-dev
```
[the remark this answers is not in the document]
Comment (yunlong-wang-cn) [reply .github/workflows/main.yml:24]: The rosdep installation and python requirement installation are currently separated in CI, therefore I think the one in rosdep won't be installed for the second task in CI, right?

File: marimbabot_speech/package.xml [hunk 16]
```diff
@@ -24,10 +23,27 @@
   <build_depend>message_generation</build_depend>
   <build_depend>sound_play</build_depend>
   <build_depend>audio_common_msgs</build_depend>
+  <build_depend>portaudio19-dev</build_depend>
```
[the remark this answers is not in the document]
Comment (yunlong-wang-cn) [reply marimbabot_speech/package.xml:26]: ok, that makes sense.

File: .github/workflows/main.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (Flova) [reply .github/workflows/main.yml:24]: The rosdep install should be first and I think since recently it is done in both CI tasks.

[the remark this answers is not in the document]
Comment (yunlong-wang-cn) [reply .github/workflows/main.yml:24]: This is what we have in the second CI task, I can not find where is the installation of rosdep. So at least not in this branch.

    build_python_requirements:
      runs-on: ubuntu-20.04
      steps:
      - uses: actions/checkout@v3
      - name: Install portaudio system dependency
        run: |
          sudo apt-get update
          sudo apt-get install -y portaudio19-dev
      - name: Set up Python 3.8.10
        uses: actions/setup-python@v4
        with:
          python-version: 3.8.10
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install wheel
          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi

[the remark this answers is not in the document]
Comment (Flova) [reply .github/workflows/main.yml:24]: Ahh nvm I somehow missunderstood the description of #253 as it only manually added one dependency. Personally I would like to include the python deps into the main CI task, but this might be a separate topic.

File: marimbabot_speech/src/speech2text.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Juphex) [reply marimbabot_speech/src/speech2text.py:1]: no

Comment (Juphex) [reply marimbabot_speech/src/speech2text.py:1]: no

**Review by Juphex — APPROVED** [review]
works as discussed
